service/controller/resource/subnet/resource.go

```diff
@@ -187,6 +212,87 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 	return nil
 }
 
+// garbageCollectSubnets removes subnets that have an ARM deployment in Azure but are not defined in `AzureCluster`.
+// This is required because when removing a node pool, we remove the subnet from `AzureCluster`, so we can remove it here from Azure.
+func (r *Resource) garbageCollectSubnets(ctx context.Context, deploymentsClient *azureresource.DeploymentsClient, subnetsClient *network.SubnetsClient, azureCluster capzv1alpha3.AzureCluster) error {
+	subnetsIterator, err := subnetsClient.ListComplete(ctx, key.ClusterID(&azureCluster), azureCluster.Spec.NetworkSpec.Vnet.Name)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for subnetsIterator.NotDone() {
+		subnetInAzure := subnetsIterator.Value()
+
+		if !isSubnetInAzureClusterSpec(ctx, azureCluster, *subnetInAzure.Name) && !isProtectedSubnet(*subnetInAzure.Name) {
+			err = r.deleteSubnet(ctx, subnetsClient, key.ClusterID(&azureCluster), azureCluster.Spec.NetworkSpec.Vnet.Name, *subnetInAzure.Name)
+			if IsSubnetInUse(err) {
```

Comment:
due to the asynchronous nature of the operators, it's possible that we try to remove the subnet even though the VMSS is still there. I don't want to show an error since this is expected. Instead, I create an error matcher and when that happens we just wait for the next reconciliation loop.